astropy/modeling/tests/test_quantities_parameters.py

```diff
@@ -0,0 +1,476 @@
+    """
+
+    class TestModel(BaseTestModel):
+        a = Parameter(default=default, unit=unit)
+
+    # TODO: decide whether the default property should return a value or
+    # a quantity?
+
+    # The default unit and value should be set on the class
+    assert TestModel.a.unit == u.m
+    assert TestModel.a.default == 1.0
+
+    # Check that the default unit and value are also set on a class instance
+    m = TestModel()
+    assert m.a.unit == u.m
+    assert m.a.default == m.a.value == 1.0
+
+    # If the parameter is set to a different value, the default is still the
+    # internal default
+    m = TestModel(2.0 * u.m)
+    assert m.a.unit == u.m
+    assert m.a.value == 2.0
+    assert m.a.default == 1.0
+
+    # Instantiate with a different, but compatible unit
+    m = TestModel(2.0 * u.pc)
+    assert m.a.unit == u.pc
+    assert m.a.value == 2.0
+    # The default is still in the original units
+    # TODO: but how do we know what those units are if we don't return a
+    # quantity?
+    assert m.a.default == 1.0
+
+    # Instantiating with incompatible units raises an error
+    with pytest.raises(InputParameterError) as exc:
+        TestModel(1.0 * u.Jy)
+    assert exc.value.args[0] == ('TestModel.__init__() requires parameter \'a\' '
+                                 'to be in units equivalent to Unit("m") '
+                                 '(got Unit("Jy"))')
+
+
+def test_parameter_quantity_arithmetic():
+    """
+    Test that arithmetic operations with properties that have units return the
+    appropriate Quantities.
+    """
+
+    g = Gaussian1D(1 * u.J, 1 * u.m, 0.1 * u.m)
+
+    # Addition should work if units are compatible
+    assert g.mean + (1 * u.m) == 2 * u.m
+    assert (1 * u.m) + g.mean == 2 * u.m
+
+
+    # Multiplication by a scalar should also preserve the quantity-ness
+    assert g.mean * 2 == (2 * u.m)
+    assert 2 * g.mean == (2 * u.m)
+
+    # Multiplication by a quantity should result in units being multiplied
+    assert g.mean * (2 * u.m) == (2 * (u.m ** 2))
+    assert (2 * u.m) * g.mean == (2 * (u.m ** 2))
+
+    # Negation should work properly too
+    assert -g.mean == (-1 * u.m)
+    assert abs(-g.mean) == g.mean
+
+    # However, addition of a quantity + scalar should not work
+    with pytest.raises(UnitsError) as exc:
+        g.mean + 1
+    assert exc.value.args[0] == ("Can only apply 'add' function to "
+                                 "dimensionless quantities when other argument "
+                                 "is not a quantity (unless the latter is all "
+                                 "zero/infinity/nan)")
+    with pytest.raises(UnitsError) as exc:
+        1 + g.mean
+    assert exc.value.args[0] == ("Can only apply 'add' function to "
+                                 "dimensionless quantities when other argument "
+                                 "is not a quantity (unless the latter is all "
+                                 "zero/infinity/nan)")
+
```

Comment:
``g.mean`` is not a quantity but a ``Parameter`` instance, so the code flow is different, and is still worth testing (``g.mean.quantity`` would be a Quantity, and in that case I agree it wouldn't be needed)